katz/src/main/kotlin/katz/typeclasses/Monad.kt

```diff
@@ -0,0 +1,50 @@
+package katz.typeclasses
+
+import java.io.Serializable
+import kotlin.coroutines.experimental.Continuation
+import kotlin.coroutines.experimental.EmptyCoroutineContext
+import kotlin.coroutines.experimental.RestrictsSuspension
+import kotlin.coroutines.experimental.intrinsics.COROUTINE_SUSPENDED
+import kotlin.coroutines.experimental.intrinsics.suspendCoroutineOrReturn
+
+interface Monad<F> : Applicative<F> {
+
+    fun <A, B> flatMap(fa: HK<F, A>, f: (A) -> HK<F, B>): HK<F, B>
+
+    override fun <A, B> ap(fa: HK<F, A>, ff: HK<F, (A) -> B>): HK<F, B> =
+            flatMap(ff, { f -> map(fa, f) })
+}
+
+@RestrictsSuspension
+open class MonadContinuation<F, A>(val M : Monad<F>) : Serializable, Continuation<HK<F, A>> {
+
```

Comment:
convention in FP instances of abstract types are usually capitalized. Also as for the top level vs inside companion we want a consistent API that is predictable when obtaining instances.  I think it's cleanr to say `Option.monad()` than `OptionMonad()` but I'm fine with whatever